Step2

hjlee83 opened the pull request:
step2 pr 요청드립니다.

hojinDev committed:
feat(dsl): add custom DSL for building person objects

Person DSL supports dynamic definition of name, company, skills, and languages.
Introduced flexible builders to simplify creation and testing of `Person` objects.

Added comprehensive tests to validate the DSL functionality and ensure correctness.

hyogeonkim committed:
feat(dsl): add language block

hyogeonkim committed:
doc: README.md

hyogeonkim committed:
feat: input playerNames

hyogeonkim committed:
feat: Deck, Card

hyogeonkim committed:
etc: formatting

hojinDev committed:
feat(readme): update checklist for blackjack requirements

aligned checklist with current implementation status by marking completed items
such as card shuffle and initial distribution. clarified pending tasks for better planning
and progress tracking.

hojinDev committed:
feat(readme): update checklist for blackjack requirements

aligned checklist with current implementation status by marking completed items
such as card shuffle and initial distribution. clarified pending tasks for better planning
and progress tracking.

hojinDev committed:
feat(blackjack): implement dealer and player mechanics with tests

added `Dealer` and `Player` classes inheriting from `Participant`, encapsulating
core blackjack logic. implemented dealing decisions for dealer based on
card value conditions. updated `Rank` enum to expose `score` for calculations.

introduced `OutView` to display game states such as initial distribution
and player card lists. added tests for `Dealer` and `Player` to validate
functionality.

updated README checklist to reflect progress on completed tasks.

hojinDev committed:
feat(blackjack): implement dealer and player mechanics with tests

added `Dealer` and `Player` classes inheriting from `Participant`, encapsulating
core blackjack logic. implemented dealing decisions for dealer based on
card value conditions. updated `Rank` enum to expose `score` for calculations.

introduced `OutView` to display game states such as initial distribution
and player card lists. added tests for `Dealer` and `Player` to validate
functionality.

updated README checklist to reflect progress on completed tasks.

hojinDev committed:
feat(blackjack): calculate ace as 1 or 11 and refactor score logic

added `totalScore` method in `Participant` to handle flexible scoring for Ace as 1 or 11.
updated `Dealer` to utilize `totalScore` for cleaner logic and improved readability.

adjusted README checklist to reflect progress on Ace scoring and dealer card addition tasks.

hojinDev committed:
feat(blackjack): calculate ace as 1 or 11 and refactor score logic

added `totalScore` method in `Participant` to handle flexible scoring for Ace as 1 or 11.
updated `Dealer` to utilize `totalScore` for cleaner logic and improved readability.

adjusted README checklist to reflect progress on Ace scoring and dealer card addition tasks.

hojinDev committed:
feat(blackjack): enhance rank display and input validation

updated `Rank` enum to include display name for improved user readability.
enhanced `InputView` to ensure all player names are non-blank during input validation.

adjusted README to refine checklist formatting and align with recent updates.

hojinDev committed:
feat(blackjack): enhance rank display and input validation

updated `Rank` enum to include display name for improved user readability.
enhanced `InputView` to ensure all player names are non-blank during input validation.

adjusted README to refine checklist formatting and align with recent updates.

hojinDev committed:
feat(blackjack): improve game flow and output readability

renamed `OutView` to `OutputView` for consistent naming convention. added `toString` method in `Card` for better card display. introduced dealer's initial card visibility and player's decision functionality for additional cards.

updated `InputView` for flexible player input handling and validation. refined exception handling by displaying error messages for invalid inputs.

hojinDev committed:
feat(blackjack): improve game flow and output readability

renamed `OutView` to `OutputView` for consistent naming convention. added `toString` method in `Card` for better card display. introduced dealer's initial card visibility and player's decision functionality for additional cards.

updated `InputView` for flexible player input handling and validation. refined exception handling by displaying error messages for invalid inputs.

hyogeonkim committed:
feat: result score, record

hojinDev committed:
Merge remote-tracking branch 'men7627/step1' into step1

# Conflicts:
#	README.md
#	src/main/kotlin/blackjack/BlackjackApplication.kt
#	src/main/kotlin/blackjack/domain/Dealer.kt
#	src/main/kotlin/blackjack/domain/Participant.kt
#	src/main/kotlin/blackjack/domain/Player.kt
#	src/main/kotlin/blackjack/model/Card.kt
#	src/main/kotlin/blackjack/model/CardEnums.kt
#	src/main/kotlin/blackjack/view/InputView.kt
#	src/main/kotlin/blackjack/view/OutputView.kt
#	src/test/kotlin/blackjack/domain/DealerTest.kt
#	src/test/kotlin/blackjack/domain/PlayerTest.kt

hojinDev committed:
feat(blackjack): enhance deck functionality and improve input handling

updated `Deck` to use a mutable list for cards, enabling proper card removal and shuffle.
added exception handling when attempting to pop cards from an empty deck. improved unit tests
to validate deck behavior, ensuring uniqueness and proper handling of an empty deck.

refined `InputView` to utilize `retryInput` for better user input validation and error handling.

hojinDev committed:
refactor(blackjack): extract reusable methods and simplify logic

replace inline score calculation and ACE check with `score()` and `isAce()` methods in `Card`.
refactor `Participant` to utilize these methods for cleaner and more readable scoring logic.

simplify `players()` function in `BlackjackApplication` by removing redundant intermediate variable.
update tests to use `toString()` for unique identification of cards, ensuring consistent behavior.

hojinDev committed:
refactor(blackjack): simplify `Dealer` additional deal condition logic

replaced block body syntax in `isAdditionalDealCondition` with expression body for brevity.
removed redundant parentheses in the `Dealer` class constructor for cleaner code.

hojinDev committed:
refactor(blackjack): move constants to companion objects for better encapsulation

relocated `DEALER_NAME`, `DEALER_ADD_CARD_BASE_SCORE`, and `ACE_ADDITIONAL_SCORE` constants from `Constants.kt` to respective companion objects in `Dealer` and `Participant`.

improves encapsulation and makes the constants more contextually relevant to their usages.

hojinDev committed:
refactor(blackjack): simplify enums and scoring logic in `Player`

reduce boilerplate in `Suit`, `Rank`, and `RecordType` enums by converting to expression bodies. update `Player` scoring method to streamline logic and enhance readability.

improves code clarity, reduces redundancy, and aligns with previous refactoring patterns.

hojinDev committed:
refactor(blackjack): enforce dealer card limit and refactor card count logic

add card limit enforcement in `Dealer`, restricting cards to a maximum of 3. introduce `cardCount()` method in `Participant` to simplify card counting logic. update relevant test cases to validate dealer's card limit and improve test descriptions for clarity.

ensures adherence to game rules by preventing dealers from exceeding allowed card limits and streamlines card-related operations across participants.

hojinDev committed:
feat(blackjack): complete game result and player victory logic

README 업데이트로 참가자 결과 출력 및 승패 출력 작업 완료 상태 반영.
`InputView`에서 `retryInput` 사용을 통해 입력 검증 및 올바른

hojinDev committed:
feat(blackjack): implement game result and improve output readability

refactor `OutputView` to streamline message generation for game results and add new methods for final output, improving readability. introduce `Result`, `PlayerRecord`, and `DealerWinDrawLose` for structured handling of match outcomes.

enhances clarity and modularity for result calculation and final display by abstracting repeated logic and providing detailed information on dealer and players.

hojinDev committed:
refactor(blackjack): ensure immutability in `cards` property of `Participant`

replace direct `_cards` reference with `.toSet()` for getter logic in `cards` property, enforcing immutability.

aligns with best practices by preventing external modifications to internal card state.

hojinDev committed:
feat(blackjack): add betting functionality and update README outputs

게임 시작 시 플레이어의 배팅 금액 설정 및 출력 기능 추가.
README에 배팅 금액 관련 입력/출력 사항 및 참가자별 수익 계산 내용 반영.

게임 진행 중 배팅 금액을 기준으로 수익/손익 계산이 가능하도록 변경, 기존 승패

hojinDev committed:
feat(blackjack): implement betting system and profit calculation

플레이어가 게임 시작 시 배팅 금액을 설정할 수 있게 기능 추가. 배팅 금액을 기반으로 손익을 계산하도록 게임 로직 수정.

딜러와 플레이어의 수익 결과를 출력하도록 변경. `RecordType`에 배당률 추가, 손익 계산 로직 간소화. 기존 승패 중심의 게임 결과 처리에서 수익 중심으로 변경.

hojinDev committed:
refactor(blackjack): extract `Hand` class and streamline participant logic

`Participant`와 관련된 카드 및 점수 계산 로직을 분리하여 `Hand` 클래스로 추출. 카드 상태 관리 및 점수 계산을 `Hand`에서 전담하도록 변경.

`Participant`와 `Dealer`, `Player`의 복잡도를 낮추고 코드의 재사용성을 높임. 관련 테스트 코드도 변경된 로직에 맞게 수정.

hojinDev committed:
refactor(blackjack): remove redundant dealer card limit test case

`Dealer` 카드 3장 제한에 대한 테스트 케이스를 삭제. 해당 로직은 이미 `Dealer` 클래스에서 처리되며, 중복된 검증이므로 필요 없음.

테스트 코드 간소화로 유지보수성을 향상시키고 불필요한 코드 제거.

hjlee83 committed:
Merge branch 'hjlee83' into step2

HyunGyu-Lee commented on src/main/kotlin/blackjack/domain/BetMoney.kt:
```diff
@@ -0,0 +1,9 @@
+package blackjack.domain
+
+class BetMoney(val money: Int) {
```
이 부분은 value class 를 활용해 볼 수있을 것 같습니다!

hjlee83 replied:
오 넵 변경했습니다.

HyunGyu-Lee commented on src/main/kotlin/blackjack/domain/Participant.kt:
```diff
@@ -1,31 +1,13 @@
-    companion object {
-        const val ACE_ADDITIONAL_SCORE = 10
-    }
+    fun isBlackjack(): Boolean = hand.count() == BLACKJACK_CARD_COUNT && score() == BLACKJACK_SCORE
```
생각차이일수도 있는데 블랙잭 여부의 경우 저는 플레이어 한정으로 쓰이는 정보라고 생각해서 Player 쪽에 두었던 것 같습니다! 근데 딜러도 블랙잭일때 뭔가 규칙이 있을수있다면 상위에 두는게 맞긴할 것 같네용

hjlee83 replied:
이번에 할때 블랙잭 규칙을 찾아서 봤는데요.
플레이어와 딜러 둘다 블랙잭일때 비긴걸로 되더라고요.

HyunGyu-Lee commented on src/main/kotlin/blackjack/domain/Player.kt:
```diff
@@ -3,23 +3,29 @@ package blackjack.domain
+        val dealerScore = dealer.score()
+
+        return when {
+            isBlackjack() && dealer.isBlackjack() -> RecordType.DRAW // 둘다 블랙잭인경우 비김
```
둘 다 블랙잭인 경우를 따로 표현해주셨군요! 요렇게 되면 Participant 에 있는게 맞겠네용

HyunGyu-Lee commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -13,7 +14,7 @@ object InputView {
                     ?.map { it.trim() }
 
-            require(playerNames != null && playerNames.isNotEmpty() && playerNames.all { it.isNotBlank() }) { "올바른 참가자 이름을 입력해주세요." }
+            require(!playerNames.isNullOrEmpty() && playerNames.all { it.isNotBlank() }) { "올바른 참가자 이름을 입력해주세요." }
```
!playerNames.isNullOrEmpty() -> playerNames.isNotEmpty() 로 대체가 가능할 것 같습니다!

```
public inline fun <T> Collection<T>.isNotEmpty(): Boolean = !isEmpty()
```

hjlee83 replied:
앞에 부정이 있어서 null 이 아니여야하는데요. playerNames.isNotEmpty()  이렇게 하면 playerNames 가 null 인경우를 거르지 못하지 않나요?

HyunGyu-Lee replied:
null 포함해서 체크하는 것을 커버하지 못하는군용.. 지금 구현 유지가 맞는것같네요!

HyunGyu-Lee commented on src/main/kotlin/blackjack/domain/PlayerProfit.kt:
```diff
@@ -0,0 +1,6 @@
+package blackjack.domain
+
+class PlayerProfit(
```
DealerProfit 과 PlayerProfit 둘 다 최종적으로 수익은 `profit` 이라는 필드로 반환되니까 이를 표현하는 상위 클래스나 인터페이스를 두면 어떨까요?!

아니면 그냥 Profit 이라고 하나만 둬도 name 으로 플레이어/딜러 구분하여 사용하면 충분히 사용은 가능할 것 같기도 하구요 (클래스가 줄어드는 대신 밖에서 계산해줘야하는 부분이 단점일수있겠네요 ㅠ)

HyunGyu-Lee reviewed (changes requested):
고생 많으셨습니다! 소소한 리뷰 몇 개 남겨두었습니다!

hojinDev committed:
refactor(blackjack): introduce `Profit` interface to unify profit handling

`Profit` 인터페이스를 도입하여 딜러와 플레이어의 수익 계산 로직을 통합. `DealerProfit`과 `PlayerProfit`이 공통 인터페이스를 구현하도록 변경.

수익 계산 구조를 단순화하고, 관련 코드의 재사용성을 향상. `OutputView` 로직 간소화 및 중복 제거를 통해 가독성을 개선.

hojinDev committed:
Merge remote-tracking branch 'origin/step2' into step2

HyunGyu-Lee reviewed (approved):
고생 많으셨습니다!! 머지할게요!
